src/backend/distributed/commands/seclabel.c

```diff
@@ -0,0 +1,165 @@
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * PreprocessSecLabelStmt is executed before the statement is applied to the local
+ * postgres instance.
+ *
+ * In this stage we can prepare the commands that need to be run on all workers to assign
+ * security labels on distributed objects, currently supporting just Role objects.
+ */
+List *
+PreprocessSecLabelStmt(Node *node, const char *queryString,
+					   ProcessUtilityContext processUtilityContext)
+{
+	if (!IsCoordinator() || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	SecLabelStmt *secLabelStmt = castNode(SecLabelStmt, node);
+
+	List *objectAddresses = GetObjectAddressListFromParseTree(node, false, false);
+	if (!IsAnyObjectDistributed(objectAddresses))
+	{
+		return NIL;
+	}
+
+	if (secLabelStmt->objtype != OBJECT_ROLE)
+	{
+		if (EnableUnsupportedFeatureMessages)
+		{
+			ereport(NOTICE, (errmsg("not propagating SECURITY LABEL commands whose"
+									" object type is not role"),
+							 errhint(
+								 "Connect to worker nodes directly to manually run the same"
+								 " SECURITY LABEL command after disabling DDL propagation.")));
+		}
+		return NIL;
+	}
+
+	if (!EnableCreateRolePropagation)
+	{
+		return NIL;
+	}
+
+	const char *sql = DeparseTreeNode((Node *) secLabelStmt);
+
+	List *commandList = list_make3(DISABLE_DDL_PROPAGATION,
+								   (void *) sql,
+								   ENABLE_DDL_PROPAGATION);
+
+	return NodeDDLTaskList(NON_COORDINATOR_NODES, commandList);
+}
+
+
+/*
+ * PostprocessSecLabelStmt ensures that all object dependencies exist on all
+ * nodes for the object in the SecLabelStmt. Currently, we only support SecLabelStmts
+ * operating on a ROLE object.
+ */
+List *
+PostprocessSecLabelStmt(Node *node, const char *queryString)
+{
+	if (!EnableCreateRolePropagation || !IsCoordinator() || !ShouldPropagate())
+	{
+		return NIL;
+	}
+
+	SecLabelStmt *secLabelStmt = castNode(SecLabelStmt, node);
+
+	if (secLabelStmt->objtype != OBJECT_ROLE)
+	{
+		return NIL;
+	}
+
+	List *objectAddresses = GetObjectAddressListFromParseTree(node, false, false);
+	if (IsAnyObjectDistributed(objectAddresses))
+	{
+		EnsureAllObjectDependenciesExistOnAllNodes(objectAddresses);
+	}
+
+	return NIL;
+}
+
+
+/*
+ * SecLabelStmtObjectAddress returns the object address of the object on
+ * which this statement operates (secLabelStmt->object). Note that it has no limitation
+ * on the object type being OBJECT_ROLE. This is intentionally implemented like this
+ * since it is fairly simple to implement and we might extend SECURITY LABEL propagation
+ * in the future to include more object types.
+ */
+List *
+SecLabelStmtObjectAddress(Node *node, bool missing_ok, bool isPostprocess)
+{
+	SecLabelStmt *secLabelStmt = castNode(SecLabelStmt, node);
+
+	Relation rel = NULL;
+	ObjectAddress address = get_object_address(secLabelStmt->objtype,
+											   secLabelStmt->object, &rel,
+											   AccessShareLock, missing_ok);
+	if (rel != NULL)
+	{
+		relation_close(rel, AccessShareLock);
+	}
```

Comment:
Can it ever return a rel when the object type is a role?

If it cannot, should replace relation_close with an ereport to make sure that this is not really called for something different than roles object yet?

Reply:
It can return a Rel in Preprocess, because there I am checking if the distributed object of the command is not a role, and if so, I give a warning. So, `GetObjectAddressListFromParseTree` in the preprocess is run for any object type.

Reply:
Is that function still supposed to process non-role objectS?

Reply:
nvm me